trieste/space.py

```diff
@@ -1060,11 +1131,136 @@ def product(self, other: TaggedProductSearchSpace) -> TaggedProductSearchSpace:
+            1,
+            message=f"""
+                All subspaces must have the same dimension but received
+                {[int(space.dimension) for space in spaces]}.
+                """,
+        )
+
+        super().__init__(spaces, tags)
+
+    @property
+    @check_shapes("return: [V, D]")
+    def lower(self) -> TensorType:
+        """Returns the stacked lower bounds of all the subspaces.
+
+        :return: The lower bounds of shape [V, D], where V is the number of subspaces and D is
+            the dimensionality of each subspace.
         """
-        if not isinstance(other, TaggedProductSearchSpace):
-            return NotImplemented
-        return self._tags == other._tags and self._spaces == other._spaces
+
+        lower = self.subspace_lower
+        return tf.stack(lower, axis=0) if lower else tf.constant([], dtype=DEFAULT_DTYPE)
+
+    @property
+    @check_shapes("return: [V, D]")
+    def upper(self) -> TensorType:
+        """Returns the stacked upper bounds of all the subspaces.
+
+        :return: The upper bounds of shape [V, D], where V is the number of subspaces and D is
+            the dimensionality of each subspace.
+        """
+
+        upper = self.subspace_upper
+        return tf.stack(upper, axis=0) if upper else tf.constant([], dtype=DEFAULT_DTYPE)
+
+    @property
+    @check_shapes("return: []")
+    def dimension(self) -> TensorType:
+        """The number of inputs in this search space."""
+        return self.get_subspace(self.subspace_tags[0]).dimension
+
+    @check_shapes("return: [num_samples, V, D]")
+    def sample(self, num_samples: int, seed: Optional[int] = None) -> TensorType:
+        """
+        Sample randomly from the space by sampling from each subspace
+        and returning the resulting samples stacked along the second axis in the same order as
+        specified when initializing the space.
+
+        :param num_samples: The number of points to sample from each subspace.
+        :param seed: Optional tf.random seed.
+        :return: ``num_samples`` i.i.d. random points, sampled uniformly,
+            from each search subspace with shape '[num_samples, V, D]' , where V is the number of
+            subspaces and D is the search space dimension.
+        """
+        return tf.stack(self.subspace_sample(num_samples, seed), axis=1)
+
+    def _contains(self, value: TensorType) -> TensorType:
+        """
+        Return `True` if ``value`` is a member of this search space, else `False`. A point
+        is a member if it is a member of any of the subspaces.
+
+        :param value: A point or points to check for membership of this :class:`SearchSpace`.
+        :return: A boolean array showing membership for each point in value.
+        """
+        return tf.reduce_any(
+            [self.get_subspace(tag)._contains(value) for tag in self.subspace_tags], axis=0
+        )
+
+    def product(self, other: TaggedMultiSearchSpace) -> TaggedMultiSearchSpace:
```

Comment:
Are there unit tests for contains, product and discretize? (and if it turns out that any of the methods don't make sense then it might be better to raise NotImplementedError rather than have untested code)

Reply:
Yes there are unit tests for all of those. E.g.
- test_collection_space_contains_point
- test_multi_space___mul___
- test_collection_space_discretize_returns_search_space_with_only_points_contained_within_box
- ... plus others